### PR TITLE
Return Not Found on /events when there is no form

### PR DESCRIPTION
**Pull request by martha**
## Description

Include a summary of the changes and a link to the related issue. List any dependencies that are required for this change.

[//]: # 'remove if not applicable'
GH issue: https://github.com/open-path/Green-River/issues/6852; comes out of: https://github.com/open-path/Green-River/issues/6793#issuecomment-2412096127 QA comment

I've targeted release 138 since this issue isn't newly introduced by release 137. (Well, it is newly exposed by adding the published scope on 137, but I guess I am trying to say that 137 is still a strict improvement, not a regression.) However, if it's easier to keep track of QA let's retarget to 137!

To reproduce:
- Temporarily disable any universal rules enabling CE Event in all projects
- Create a new CE Event form, but don't publish it - leave it in draft
- Add an active Rule enabling this new CE Event form in a project
- Navigate to an enrollment dashboard in the project. While you cannot see the CE Event link in the left-hand nav, if you directly type in /events to the url, you can still access the CE Events page and click the button to add an event.
  - Example provided by Angela in QA: https://qa-hmis.openpath.host/client/==VjlBUnZjbzhWaEtHYTg0S2Z2UVJJQT09/enrollments/==WHpNUG9OS0Y3Z2R0dGZxdTN1WnBUdz09/events
- After this PR, you should see "Not Found" instead of being able to access the page

## Type of change
[//]: # 'remove options that are not relevant'
- [x] Bug fix
- [ ] New feature (adds functionality)
- [ ] Code clean-up / housekeeping
- [ ] Dependency update

## Checklist before requesting review
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] My code includes comments and/or descriptive variable names to help other engineers understand the intent (or not applicable)
- [x] My code follows the style guidelines of this project (eslint)
- [n/a] I have updated the documentation (or not applicable)
- [x] If it's not obvious how to test this change, I have provided testing instructions in this PR or the related issue

### CODE REVIEW

File: src/modules/enrollment/components/dashboardPages/EnrollmentCeEventsPage.tsx [hunk 1]
```diff
@@ -62,21 +62,27 @@ const EnrollmentCeEventsPage = () => {
...
-      localConstants,
-      projectId: enrollment?.project.id,
-    });
+  const {
+    onSelectRecord,
+    viewRecordDialog,
+    editRecordDialog,
+    openFormDialog,
+    formNotFound,
+  } = useViewEditRecordDialogs({
```
Comment (gigxz) [inline src/modules/enrollment/components/dashboardPages/EnrollmentCeEventsPage.tsx]: I wouldn't think this is specific to CE Events, is it? There are plenty of other pages that use this pattern: I'm seeing CE Assessments, CLS, Case Notes, HMIS Participation, and CE Participation.

This approach feels a little tricky to maintain, because each component that uses the `useViewEditRecordDialogs` hook needs to realize that it's possible for the hook to fail, and that the component should look at `formNotFound` and render a NotFound component. The fact that the form isn't found feels like something that shouldn't be the concern of the dialog hook. IMO would make more sense if the dialog hook just throw an error if the form wasn't found–– because if the form wasn't found, it cannot render a dialog safely.

I'd propose that we use the same logic that we already use to determine whether the given page shows up in the left nav. For CE Events, that's:
https://github.com/greenriver/hmis-frontend/blob/84fccbcbedc112da322fe8f855f0ea379dccd73a/src/modules/enrollment/hooks/useEnrollmentDashboardNavItems.tsx#L50-L52

I think it would be ideal if the app could say "you're trying to navigate to `/events` but events wasn't included in the nav items, so I'm going to return NotFound [or redirect you to the enrollment dash overview]" without bothering to attempt to load the page itself. If we can do that in a way that is extensible to all dashboard pages, we don't need to worry about it in each pages render logic. (And we can raise an error if the form isn't found, because we don't expect that)


We do already have the enabledFeatures list in the context:
```
const { enabledFeatures } = useEnrollmentDashboardContext();
  if (!enabledFeatures.includes(DataCollectionFeatureRole.CeEvent)) {
    return <NotFound />;
  }
```

Looking at the `protected.tsx` routes, I wonder if we can just introduce a prop to `EnrollmentRoute` to this effect:
```
            element: (
              <EnrollmentRoute
                dataCollectionFeature={DataCollectionFeatureRole.CeEvent|
                redirectRoute={EnrollmentDashboardRoutes.ENROLLMENT_OVERVIEW}
              >
                <EnrollmentCeEventsPage />
              </EnrollmentRoute>
            ),
```
Similar to how `EnrollmentRoute` handles returning NotFound (or redirect) if the user lacks permission, it could likewise return NotFound (or redirect) if the Data Collection Feature is not listed in the enrollment dashboard contexts' `enabledFeatures` list.

Comment (martha) [reply src/modules/enrollment/components/dashboardPages/EnrollmentCeEventsPage.tsx]: Thanks @gigxz  I like that idea, it's really elegant. 

Do you think we should implement something similar in `ProjectRoute`, for HMIS Participation, CE Participation, Current living situation, and maybe external form submissions?

File: src/modules/form/hooks/useFormDialog.tsx [hunk 2]
```diff
@@ -83,7 +83,12 @@ export function useFormDialog<T extends SubmitFormAllowedTypes>({
       projectId:
         projectId || localConstants?.projectId || inputVariables?.projectId,
     },
-    localDefinition
+    localDefinition,
```
Comment (martha) [inline src/modules/form/hooks/useFormDialog.tsx]: I was just fiddling around here, but maybe we should just revert this change. Throwing the error here inside of the hook (as opposed to in a rendered component) just errors in the console. But it will only get thrown visibly to the user when they try to actually render a component that uses the form definition, like on line 130 below.

File: src/routes/protected.tsx [hunk 3]
```diff
@@ -485,13 +486,27 @@ export const protectedRoutes: RouteNode[] = [
           },
           {
             path: EnrollmentDashboardRoutes.CURRENT_LIVING_SITUATIONS,
-            // No perm needed because it only requires enrollment visibility
-            element: <EnrollmentCurrentLivingSituationsPage />,
+            element: (
+              <EnrollmentRoute
+                permissions={[]} // Only requires enrollment visibility
```
Comment (martha) [inline src/routes/protected.tsx]: Could pass `canViewEnrollments`?

File: src/modules/form/hooks/useFormDialog.tsx (hunk 2 above)
Comment (gigxz) [reply src/modules/form/hooks/useFormDialog.tsx]: I'm not sure I see the benefit to throwing the error from a callback vs throwing it in the component. Thanks to typescript, I think most of the components that call `useDefinition` already _have to_ deal with the state where `!loading && !formDefinition`--  most of them error in this case, but some render a NotFound component, which I think is wrong (`ViewRecord`, `EditRecord`). In the recent issue that came up-- where a user saw "Not Found" in the case note modal-- it would have been preferable if that had been treated as an error and reported through sentry so we could catch it asap.

I _think_ it would make sense for `useFormDefinition` to just raise an error if the fetch turns up empty. I can't think of any cases where we would expect that, but maybe I'm missing something.

File: src/modules/enrollment/components/dashboardPages/EnrollmentCeEventsPage.tsx (hunk 1 above)
Comment (gigxz) [reply src/modules/enrollment/components/dashboardPages/EnrollmentCeEventsPage.tsx]: Yes I think it would make sense to have in the project routes as well, since it uses the same concept of data collection features (For Services and Current Living Situations). I don't think the same would apply to HMIS Participation/ CE participation since those aren't "optional" features.

File: src/routes/protected.tsx (hunk 3 above)
Comment (martha) [reply src/routes/protected.tsx]: We can't pass an empty array, since it makes the `hasPermissions` check evaluate as false. I made `permissions` optional on the `EnrollmentRoute` and `ProjectRoute` instead.

File: src/modules/form/hooks/useFormDialog.tsx (hunk 2 above)
Comment (martha) [reply src/modules/form/hooks/useFormDialog.tsx]: Sounds good, I added the throw in `useFormDefinition`. 

I also removed the throw/NotFound returns from the other components, since I thought it's a bit clearer if this situation is only explicitly dealt with in one place. But I wonder if this is a bit messy. I want to think about it more.

Comment (gigxz) [reply src/modules/form/hooks/useFormDialog.tsx]: Hmm I think I see what you mean. I would look at `EditRecord` for example and say "oh, so this component will render null if there is no form" (since it has that `formDefinition && ` conditional render syntax). Which of course isn't true, because that won't get hit. I guess I would almost prefer if BOTH threw an error, so in EditRecord have `if (!formDefinition) throw "form definition not found"` to appease typescript, since at least we're not implying something different with the render. That error wont get hit, of course, because useFormDefinition will have already raised.
I also REALLY like that the query variables are included when raising the error from useFormDefinition, so that will help us track it down. I think that's another "plus" to raising from useFormDefinition.

Comment (martha) [reply src/modules/form/hooks/useFormDialog.tsx]: Ok, this makes sense. I agree with you, the most straightforward thing is for the error to get raised in several places so that we are never confused looking at the code about how it will deal with an absent form definition!

File: src/components/accessWrappers/ProjectRoute.tsx [hunk 4]
```diff
@@ -14,19 +15,29 @@ import { generateSafePath } from '@/utils/pathEncoding';
...
     redirectRoute?: string;
+    dataCollectionFeature?: DataCollectionFeatureRole;
   }>
-> = ({ permissions, redirectRoute, children }) => {
+> = ({ permissions, redirectRoute, dataCollectionFeature, children }) => {
   const { project } = useProjectDashboardContext();
   const [hasPermission, { loading }] = useHasProjectPermissions(
     project.id,
-    ensureArray(permissions)
+    permissions ? ensureArray(permissions) : []
```
Comment (gigxz) [inline src/components/accessWrappers/ProjectRoute.tsx]: `ensureArray` should handle the case when permissions is null or undefined- that util is meant to help us avoid ternaries like this!

**Review by gigxz — APPROVED** [review]
thanks, looks good to me!

File: src/modules/form/hooks/useFormDialog.tsx (hunk 2 above)
Comment (gigxz) [reply src/modules/form/hooks/useFormDialog.tsx]: okay, sounds good

File: src/routes/protected.tsx (hunk 3 above)
Comment (gigxz) [reply src/routes/protected.tsx]: Ok, sounds good

File: src/components/accessWrappers/EnrollmentRoute.tsx [hunk 5]
```diff
@@ -5,21 +5,34 @@ import NotFound from '../pages/NotFound';
...
+  );
+
+  if (
+    dataCollectionFeature &&
+    !enabledFeatures.includes(dataCollectionFeature)
+  ) {
+    return <NotFound />;
+  }
+
+  if (!!permissions && !hasPermission) {
```
Comment (gigxz) [inline src/components/accessWrappers/EnrollmentRoute.tsx]: Is it intentional that this will always return not found if you pass `permissions=[]`? We could make it more robust to handle that in the same way:
`if (ensureArray(permissions).length > 0 && !hasPermission) {`

File: src/components/accessWrappers/ProjectRoute.tsx (hunk 4 above)
Comment (martha) [reply src/components/accessWrappers/ProjectRoute.tsx]: Sounds good, `ensureArray` just needed a small typescript update to allow. https://github.com/greenriver/hmis-frontend/pull/959/files#diff-d03a2dd2397fbec94efa7e9b0cb974eada2c960a0b3c964fe6b4822c9e809fbbR1

File: src/components/accessWrappers/EnrollmentRoute.tsx (hunk 5 above)
Comment (martha) [reply src/components/accessWrappers/EnrollmentRoute.tsx]: I agree with your suggestion and made the change to make it more robust. With this, we could also now revert the change I mentioned [here](https://github.com/greenriver/hmis-frontend/pull/959#discussion_r1804664499) making `permissions` an optional param. But I kind of like that change anyway, it seems like a clearer api for the Route components now that they are not only used for checking permissions
